### PR TITLE
[GarbageCollector] let dynamic client handle non-registered ListOptions

### LINKED ISSUE #27622: Garbage collection integration tests produce prolific amount of message spam
It seems to me like all of the garbage collection integration tests produce a ton of log messages of the form:

```
E0617 11:41:53.958594   26645 reflector.go:216] pkg/controller/garbagecollector/garbagecollector.go:690: Failed to list <nil>: no kind "ListOptions" is registered for version "policy/v1alpha1"
E0617 11:41:54.958807   26645 reflector.go:216] pkg/controller/garbagecollector/garbagecollector.go:690: Failed to list <nil>: no kind "ListOptions" is registered for version "policy/v1alpha1"
E0617 11:41:55.959029   26645 reflector.go:216] pkg/controller/garbagecollector/garbagecollector.go:690: Failed to list <nil>: no kind "ListOptions" is registered for version "policy/v1alpha1"
E0617 11:41:56.959258   26645 reflector.go:216] pkg/controller/garbagecollector/garbagecollector.go:690: Failed to list <nil>: no kind "ListOptions" is registered for version "policy/v1alpha1"
```

### PR DESCRIPTION
**Pull request by caesarxuchao**
And register v1.ListOptions in the policy group.

Fix #27622

@lavalamp @smarterclayton @krousey

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27733)

### CODE REVIEW

File: pkg/client/typed/dynamic/client.go [hunk 1]
```diff
@@ -270,3 +270,23 @@ func (parameterCodec) DecodeParameters(parameters url.Values, from unversioned.G
 }
 
 var defaultParameterEncoder runtime.ParameterCodec = parameterCodec{}
+
+type versionedParameterEncoderWithV1Fallback struct{}
```
Comment (caesarxuchao) [inline pkg/client/typed/dynamic/client.go]: Registering v1.ListOptions in the Policy group will fix #27622. But for thirdparty resources, we still need this versionedParameterEncoderWithV1Fallback codec.

For the record, API server always treats query parameters sent to a thirdparty resource endpoint as v1: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/thirdpartyresourcedata/codec.go#L557

File: pkg/apis/policy/v1alpha1/register.go [hunk 2]
```diff
@@ -41,6 +42,7 @@ func addKnownTypes(scheme *runtime.Scheme) {
 	scheme.AddKnownTypes(SchemeGroupVersion,
 		&PodDisruptionBudget{},
 		&PodDisruptionBudgetList{},
+		&v1.ListOptions{},
```
Comment (smarterclayton) [inline pkg/apis/policy/v1alpha1/register.go]: What about DeleteOptions?  And ExportOptions?

File: pkg/client/typed/dynamic/client.go [hunk 3]
```diff
@@ -270,3 +270,23 @@ func (parameterCodec) DecodeParameters(parameters url.Values, from unversioned.G
...
+
+func (versionedParameterEncoderWithV1Fallback) EncodeParameters(obj runtime.Object, to unversioned.GroupVersion) (url.Values, error) {
+	if ret, err := api.ParameterCodec.EncodeParameters(obj, to); err != nil && runtime.IsNotRegisteredError(err) {
+		// fallback to v1
+		return api.ParameterCodec.EncodeParameters(obj, v1.SchemeGroupVersion)
+	} else {
```
Comment (smarterclayton) [inline pkg/client/typed/dynamic/client.go]: style: we prefer

```
ret, err := ....
if err != nil && ... {
  return ...
}
return ret, err
```

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 4]
```diff
@@ -443,21 +443,15 @@ func gcListWatcher(client *dynamic.Client, resource unversioned.GroupVersionReso
...
 		WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
 			// APIResource.Kind is not used by the dynamic client, so
 			// leave it empty. We want to list this resource in all
 			// namespaces if it's namespace scoped, so leave
 			// APIResource.Namespaced as false is all right.
 			apiResource := unversioned.APIResource{Name: resource.Resource}
-			// The default parameter codec used by the dynamic client cannot
-			// encode api.ListOptions.
-			return client.ParameterCodec(api.ParameterCodec).Resource(&apiResource, api.NamespaceAll).Watch(&options)
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).Resource(&apiResource, api.NamespaceAll).Watch(&options)
```
Comment (smarterclayton) [inline pkg/controller/garbagecollector/garbagecollector.go]: Isn't the namespace controller also going to hit this?

Comment (smarterclayton) [inline pkg/controller/garbagecollector/garbagecollector.go]: @derekwaynecarr and also the quota controller?

File: pkg/client/typed/dynamic/client.go [hunk 5]
```diff
@@ -270,3 +270,23 @@ func (parameterCodec) DecodeParameters(parameters url.Values, from unversioned.G
...
+		return ret, err
+	}
+}
+func (versionedParameterEncoderWithV1Fallback) DecodeParameters(parameters url.Values, from unversioned.GroupVersion, into runtime.Object) error {
```
Comment (smarterclayton) [inline pkg/client/typed/dynamic/client.go]: extra line and godoc

File: pkg/api/serialization_test.go [hunk 6]
```diff
@@ -174,6 +174,16 @@ var nonRoundTrippableTypes = sets.NewString(
 	"WatchEvent",
 )
 
+// verify all external group/versions have ListOptions registered.
+func TestListOptionsRegistered(t *testing.T) {
+	for _, group := range testapi.Groups {
+		gv := group.GroupVersion()
+		if _, err := api.Scheme.New(gv.WithKind("ListOptions")); err != nil {
```
Comment (smarterclayton) [inline pkg/api/serialization_test.go]: Expand to catch the other types, also watch events (I guess there are a few groups not covered by that, but I think they have their own schemes).

File: pkg/apis/policy/v1alpha1/register.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apis/policy/v1alpha1/register.go]: DeleteOptions makes sense. I'm not sure about ExportOptions. It's always treated as "v1": https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/resthandler.go#L135, though that code might be wrong.

File: pkg/client/typed/dynamic/client.go (hunk 5 above)
Comment (caesarxuchao) [inline pkg/client/typed/dynamic/client.go]: Extra line added. I think our tradition is keeping the godoc in the interface so I'll omit the godoc.

File: pkg/client/typed/dynamic/client.go (hunk 3 above)
Comment (caesarxuchao) [inline pkg/client/typed/dynamic/client.go]: Thanks. Done.

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 4 above)
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector.go]: Namespace controller lists and watches namespaces only, so it's *not* using the dynamic client for listing or watching: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/namespace_controller.go#L82-L88.

Same for the quota controller: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/resourcequota/resource_quota_controller.go#L93-L99

File: pkg/apis/policy/v1alpha1/register.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apis/policy/v1alpha1/register.go]: Added DeleteOptions but not the ExportOptions.

**Comment by caesarxuchao** [comment]
@k8s-bot node e2e test this, issue #27779

File: pkg/apis/policy/v1alpha1/register.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/apis/policy/v1alpha1/register.go]: Wow, that code is wrong.

Comment (caesarxuchao) [inline pkg/apis/policy/v1alpha1/register.go]: ExportOptions is registered as unversioned type: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/register.go#L108.

We should register it as an ordinary type. @smarterclayton Are you fine if I do that in another PR? Moving ExportOptions from unversioned/ to a api/ is an API change.

Comment (smarterclayton) [inline pkg/apis/policy/v1alpha1/register.go]: We can do it in another issue - it's no less broken then (everything in an
API is versioned, even if we haven't explicitly created a second versioned
yet).  Said another way, unversioned is something we haven't correctly
versioned yet (a bug).

It's not really an API change - the schema accepted is a version, so it's
really just an internal error that each group version doesn't explicitly
say what schema of export options they support.

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 4 above)
Comment (derekwaynecarr) [inline pkg/controller/garbagecollector/garbagecollector.go]: quota does not yet use dynamic client anywhere, but it would be nice to use it where we dont have special behavior.

namespace controller does use dynamic client to list when doing cleanup:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/namespace_controller_utils.go#L198

Comment (derekwaynecarr) [inline pkg/controller/garbagecollector/garbagecollector.go]: that said, we dont do anything with the data coming back other than work with the unstructured list.

Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector.go]: Namespace controller uses `v1.ListOptions` when listing or deleting via the dynamic client:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/namespace_controller_utils.go#L198
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/namespace_controller_utils.go#L156, so it's not affected.

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 7]
```diff
@@ -443,21 +443,15 @@ func gcListWatcher(client *dynamic.Client, resource unversioned.GroupVersionReso
 			// namespaces if it's namespace scoped, so leave
 			// APIResource.Namespaced as false is all right.
 			apiResource := unversioned.APIResource{Name: resource.Resource}
-			// The default parameter codec used by the dynamic client cannot
-			// encode api.ListOptions.
-			// TODO: api.ParameterCodec doesn't support thirdparty objects.
-			// We need a generic parameter codec.
-			return client.ParameterCodec(api.ParameterCodec).Resource(&apiResource, api.NamespaceAll).List(&options)
+			return client.ParameterCodec(dynamic.VersionedParameterEncoderWithV1Fallback).Resource(&apiResource, api.NamespaceAll).List(&options)
```
Comment (lavalamp) [inline pkg/controller/garbagecollector/garbagecollector.go]: nit: line breaks make stuff like this easier to read.

**Comment by lavalamp** [comment]
LGTM-- please add the comment and apply the label yourself.

File: pkg/apis/policy/v1alpha1/register.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apis/policy/v1alpha1/register.go]: Tracked in #27886

File: pkg/client/typed/dynamic/client.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/client/typed/dynamic/client.go]: Added to the comment on `VersionedParameterEncoderWithV1Fallback`

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 7 above)
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector.go]: Done.

**Comment by caesarxuchao** [comment]
Comments addressed. Self applied the label. Thank you both for the review.

**Comment by goltermann** [comment]
@caesarxuchao looks like the build failed?  Can you take a look?

FAIL	k8s.io/kubernetes/pkg/api	24.789s

**Comment by caesarxuchao** [comment]
@goltermann thanks for the reminder!

**Comment by caesarxuchao** [comment]
Registered v1.DeleteOptions in batch/v2alpha1, that should fix the test. Reapplying the lgtm.
